trappy/ftrace.py

```diff
@@ -173,74 +173,113 @@ def _trace_cache_path(self):
         cache_path = os.path.join(tracefile_dir, cache_dir)
         return cache_path
 
-    def _check_trace_cache(self):
+    def _get_cache_metadata(self):
         cache_path = self._trace_cache_path()
-        trace_metadata_path = os.path.join(cache_path, 'metadata.json')
+        metadata_path = os.path.join(cache_path, 'metadata.json')
 
-        if not os.path.exists(trace_metadata_path):
-            return False
+        if not os.path.exists(metadata_path):
+            return {}
+
+        metadata = {}
+        with open(metadata_path) as f:
+            metadata = json.load(f)
+
+        return metadata
 
-        with open(trace_metadata_path) as f:
-            trace_metadata = json.load(f)
+    def _is_cache_valid(self, cache_metadata):
+        for key in ["md5sum", "basetime"]:
+            if key not in cache_metadata.keys():
+                warnstr = "Cache metadata is erroneous, invalidating cache"
+                warnings.warn(warnstr)
+                return False
 
         with open(self.trace_path, 'rb') as f:
             trace_md5sum = hashlib.md5(f.read()).hexdigest()
 
-        self.basetime = float(trace_metadata["basetime"])
-
-        # Check if cache is valid
-        if trace_metadata["md5sum"] != trace_md5sum:
+        if cache_metadata["md5sum"] != trace_md5sum:
             warnstr = "Cached data is from another trace, invalidating cache."
             warnings.warn(warnstr)
-            shutil.rmtree(cache_path)
             return False
 
         return True
 
-    def _create_trace_cache(self):
+    def _prepare_cache_dir(self):
         cache_path = self._trace_cache_path()
-        trace_metadata_path = os.path.join(cache_path, 'metadata.json')
 
         if os.path.exists(cache_path):
             shutil.rmtree(cache_path)
         os.mkdir(cache_path)
 
-        trace_metadata = {}
-
-        trace_metadata["md5sum"] = hashlib.md5(open(self.trace_path, 'rb').read()).hexdigest()
-        trace_metadata["basetime"] = self.basetime
-
-        with open(trace_metadata_path, 'w') as f:
-            json.dump(trace_metadata, f)
-
     def _update_cache(self):
         try:
             # Recreate basic cache directories only if nothing cached
             if not any([c.cached for c in self.trace_classes]):
-                self._create_trace_cache()
+                self._prepare_cache_dir()
+
+                # Write cache metadata
+                metadata_path = os.path.join(self._trace_cache_path(), 'metadata.json')
 
-            # Write out only events that weren't cached before
+                metadata = {}
+                metadata["md5sum"] = hashlib.md5(
+                    open(self.trace_path, 'rb').read()
+                ).hexdigest()
+                metadata["basetime"] = self.basetime
+                metadata["extra"] = self._get_extra_data_to_cache()
```

Comment:
Instead of having an "extra" key in `metadata` and `_get_extra_data_to_cache` and `_load_extra_data_from_cache`, have you considered just having `_get_metadata_to_cache` and `_load_metadata_from_cache`, then having the subclasses use `super` then extend the result?

So, `GenericFTrace._get_metadata_to_cache` just returns `{"md5sum": <hash>,  "basetime": self.basetime"}`, then `FTrace` overrides it like:

```python
def _get_metadata_to_cache(self):
    ret = super(FTrace, self)._get_metadata_to_cache()
    ret['_version'] = self._version
    ret['_cpus'] = self._cpus
    return ret
```

Then `GenericFTrace._load_metadata_from_cache` (called before `_is_cache_valid`) just sets `self.basetime`, and `FTrace` overrides it like:

```python
def _load_metadata_from_cache(self, metadata):
    ret = super(FTrace, self)._load_metadata_from_cache(metadata):
    self._cpus = metadata['_cpus']
    self._version = metadata['_version']
```

It feels more in the noble spirit of OOP, rather than drawing a line in the inheritance tree after which everything is 'extra'. Do you think that's clearer?

Reply:
PS one benefit of writing it this way is you can grep for "self._cpus =" when debugging. That's orthogonal to the inheritance design though.

Reply:
I did think about that - it would have worked the same way as `_parsing_setup` & `_parsing_teardown`.

The thing is, `file_to_parse` is set in  `_parsing_setup` but it is **meant** to be changed. So you can override it and not even call the super method, and it's fine.
You could break `_get_metadata_to_cache` by not calling the super method (though that'd just be stupid) OR by writing over pre-existing metadata values (unlikely, but I wanted to play safe). The idea of that "extra" key is to provide a "safe" dict where any key can be used without compromising the required base metadata.

Reply:
Arguably this "extra" dict could be introduced by whatever subclass is overriding this method... So if you're okay with it we could have the metadata handling be done in overridable methods as you suggested.

Reply:
As I considered your points I became more sure that `super` is better :grin: 

> You could break _get_metadata_to_cache by not calling the super method (..) OR by writing over pre-existing metadata values (..).  The idea of that "extra" key is to provide a "safe" dict where any key can be used without compromising the required base metadata.

I guess this is a question of how to do coupling between parent and child classes - should the child be aware of the parent (must call the super method and be aware of keys it shouldn't touch) or the parent be aware of the child (must define and call an "extra" method and write an "extra" key). `super` is a language feature and all child classes must be aware of their parents anyway, so I think the former is more harmonious.

On the other hand, the thing about the child having to know which keys it mustn't touch is a good point. I'm not sure how much it matters when the classes are this tightly coupled anyway, though (they live in the same file). 

Anyway, this is probably getting a bit too philosophical so I'll leave the decision to you :)

Reply:
I think that all things considered, the overriding of a base `_get/load_metadata_from/to_cache` is the less confusing way to go. I'll still use an "extra" key, but defined in the child.